### PR TITLE
Adding tidymodels support

### PR DESCRIPTION
**Pull request by dgkf**
Starts to address #169

# Description

This takes a bit of a different tack than what was described in #169, but will be functionally similar. 

`multilevelmod` registers a bunch of package functions as `tidymodels`-compatible model functions. It still requires that those modelling packages are installed separately. Instead, we can support `tidymodels` directly. I believe that the `tidymodels` team probably maintains `multilevelmod` themselves, presumably because these packages are unlikely to add `tidymodels` compatibility. We don't have that constraint and can implement it directly. Though I'm not totally keen to how that ecosystem operates, so I'm open to other options if there's some other benefit that isn't obvious to me.

This PR sets the groundwork for `parsnip` interoperability, though we should hold off until #189 is merged to add in a predict method for `tidymodels` to use.

In addition, I generalized our registration against `emmeans` so that it can be reused for `parsnip`, and also added in some functionality so that we register against those packages even if they're loaded after `mmrm`.

# See it in action

Adapting the workflow in the [intro vignette](https://openpharma.github.io/mmrm/main/articles/introduction.html), this is how the same workflow is supported now in `tidymodels`

```r
library(tidymodels)
library(mmrm)

model <- linear_reg() %>%
  set_engine("mmrm") %>%
  fit(FEV1 ~ RACE + ARMCD * AVISIT + us(AVISIT | USUBJID), fev_data)
```

### CODE REVIEW

File: R/interop-parsnip.R
Comment (dgkf) [inline R/interop-parsnip.R:69]: This is a placeholder, awaiting #189

**Comment by danielinteractive** [comment]
@yonicd would you like to review this one ?

**Comment by yonicd** [comment]
Sure. I can review this.

**Comment by danielinteractive** [comment]
@yonicd I just gave you write access to the repo, I think that should allow you to review this PR :-) thanks!!

File: R/interop-parsnip.R
Comment (dgkf) [reply R/interop-parsnip.R]: For now I just removed this and we can add back in a predict method when it's available.

File: R/interop-parsnip.R
Comment (dgkf) [inline R/interop-parsnip.R:43]: It's not clear to me how arguments registered with `parsnip::set_model_arg` are supposed to be used. From what I can tell, unless the argument maps to some `tidymodels` whitelist, it isn't used?

Maybe someone with more `tidymodels` familiarity can weigh in

Comment (danielinteractive) [reply R/interop-parsnip.R:43]: @yonicd do you know? (I'm so far not a user of that)

Comment (dgkf) [reply R/interop-parsnip.R:43]: I'll continue to dig into it today. From what I understand, `tidymodels` compartmentalizes arguments into tuning parameters and model behavior arguments.

I don't think we have anything that would be considered a tuning parameters, but I would like to expose the behavior arguments if possible.

File: R/zzz.R
Comment (danielinteractive) [reply R/zzz.R:10]: When would this message the user? When `emmeans` is loaded and `mmrm` too?

File: R/interop-parsnip.R
Comment (dgkf) [reply R/interop-parsnip.R]: Welp, turns out you don't have to register args. It just wasn't clear to me how they get introduced. After some toying around with `tidymodels` today I found that by _not_ registering arguments, they can be passed to `set_engine` :shrug: 

This will also capture `...` arguments and propagate them to `control`/`h_get_optimizers`

```r
linear_reg() %>%
  set_engine("mmrm", reml = FALSE) %>%
  fit(FEV1 ~ RACE + ARMCD * AVISIT + us(AVISIT | USUBJID), fev_data)
```

File: R/zzz.R
Comment (danielinteractive) [reply R/zzz.R:10]: Looks nice- yeah maybe fancy if we match those packages layouts

**Comment by yonicd** [comment]
got that invite today. will look at the current code tonight.

**Comment by yonicd** [comment]
Looks good overall. a few small comments that I'm sure how to get around

It is good that user can input DF method using set_engine params

``` r
library(tidymodels)
library(mmrm)
#> mmrm() registered as tidymodel linear_reg model

model <- linear_reg() |>
  set_engine("mmrm", control = mmrm_control(method = 'Satterthwaite')) |>
  fit(FEV1 ~ RACE + ARMCD * AVISIT + us(AVISIT | USUBJID), fev_data)

summary(model$fit)
#> mmrm fit
#> 
#> Formula:     FEV1 ~ RACE + ARMCD * AVISIT + us(AVISIT | USUBJID)
#> Data:        data (used 537 observations from 197 subjects with maximum 4 
#> timepoints)
#> Covariance:  unstructured (10 variance parameters)
#> Method:      Satterthwaite
#> Inference:   REML
#> 
#> Model selection criteria:
#>      AIC      BIC   logLik deviance 
#>   3407.4   3440.2  -1693.7   3387.4 
#> 
#> Coefficients: 
#>                                Estimate Std. Error        df t value Pr(>|t|)
#> (Intercept)                    30.96770    0.82933 187.91000  37.340  < 2e-16
#> RACEBlack or African American   1.50465    0.62066 169.95000   2.424   0.0164
#> RACEWhite                       5.61310    0.66309 158.87000   8.465 1.61e-14
#> ARMCDTRT                        3.77556    1.07628 146.27000   3.508   0.0006
#> AVISITVIS2                      4.82859    0.80171 143.66000   6.023 1.37e-08
#> AVISITVIS3                     10.33317    0.82244 155.66000  12.564  < 2e-16
#> AVISITVIS4                     15.05256    1.31286 138.39000  11.465  < 2e-16
#> ARMCDTRT:AVISITVIS2            -0.01737    1.12916 138.39000  -0.015   0.9877
#> ARMCDTRT:AVISITVIS3            -0.66753    1.18654 158.21000  -0.563   0.5745
#> ARMCDTRT:AVISITVIS4             0.63094    1.85079 129.64000   0.341   0.7337
#>                                  
#> (Intercept)                   ***
#> RACEBlack or African American *  
#> RACEWhite                     ***
#> ARMCDTRT                      ***
#> AVISITVIS2                    ***
#> AVISITVIS3                    ***
#> AVISITVIS4                    ***
#> ARMCDTRT:AVISITVIS2              
#> ARMCDTRT:AVISITVIS3              
#> ARMCDTRT:AVISITVIS4              
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
#> 
#> Covariance estimate:
#>         VIS1    VIS2    VIS3    VIS4
#> VIS1 40.7335 14.2740  5.1411 13.5288
#> VIS2 14.2740 26.2243  2.6391  7.3219
#> VIS3  5.1411  2.6391 14.9497  1.0341
#> VIS4 13.5288  7.3219  1.0341 95.6006

model2 <- linear_reg() |>
  set_engine("mmrm", method = 'Satterthwaite') |>
  fit(FEV1 ~ RACE + ARMCD * AVISIT + us(AVISIT | USUBJID), fev_data)

summary(model2$fit)
#> mmrm fit
#> 
#> Formula:     FEV1 ~ RACE + ARMCD * AVISIT + us(AVISIT | USUBJID)
#> Data:        data (used 537 observations from 197 subjects with maximum 4 
#> timepoints)
#> Covariance:  unstructured (10 variance parameters)
#> Method:      Satterthwaite
#> Inference:   REML
#> 
#> Model selection criteria:
#>      AIC      BIC   logLik deviance 
#>   3407.4   3440.2  -1693.7   3387.4 
#> 
#> Coefficients: 
#>                                Estimate Std. Error        df t value Pr(>|t|)
#> (Intercept)                    30.96770    0.82933 187.91000  37.340  < 2e-16
#> RACEBlack or African American   1.50465    0.62066 169.95000   2.424   0.0164
#> RACEWhite                       5.61310    0.66309 158.87000   8.465 1.61e-14
#> ARMCDTRT                        3.77556    1.07628 146.27000   3.508   0.0006
#> AVISITVIS2                      4.82859    0.80171 143.66000   6.023 1.37e-08
#> AVISITVIS3                     10.33317    0.82244 155.66000  12.564  < 2e-16
#> AVISITVIS4                     15.05256    1.31286 138.39000  11.465  < 2e-16
#> ARMCDTRT:AVISITVIS2            -0.01737    1.12916 138.39000  -0.015   0.9877
#> ARMCDTRT:AVISITVIS3            -0.66753    1.18654 158.21000  -0.563   0.5745
#> ARMCDTRT:AVISITVIS4             0.63094    1.85079 129.64000   0.341   0.7337
#>                                  
#> (Intercept)                   ***
#> RACEBlack or African American *  
#> RACEWhite                     ***
#> ARMCDTRT                      ***
#> AVISITVIS2                    ***
#> AVISITVIS3                    ***
#> AVISITVIS4                    ***
#> ARMCDTRT:AVISITVIS2              
#> ARMCDTRT:AVISITVIS3              
#> ARMCDTRT:AVISITVIS4              
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
#> 
#> Covariance estimate:
#>         VIS1    VIS2    VIS3    VIS4
#> VIS1 40.7335 14.2740  5.1411 13.5288
#> VIS2 14.2740 26.2243  2.6391  7.3219
#> VIS3  5.1411  2.6391 14.9497  1.0341
#> VIS4 13.5288  7.3219  1.0341 95.6006
```

<sup>Created on 2023-01-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

what may be a problem is that people cant use recipes because there is a function in the formula and those should be recipe steps

``` r
library(tidymodels)
library(mmrm)
#> mmrm() registered as tidymodel linear_reg model

recipe(FEV1 ~ RACE + ARMCD + AVISIT + us(AVISIT | USUBJID), data = fev_data) |>
  step_dummy(ARMCD) |>
  step_interact(terms = c('ARMCD','AVISIT'))
#> Error in `inline_check()`:
#> ! No in-line functions should be used here; use steps to define baking actions.

#> Backtrace:
#>     ▆
#>  1. ├─recipes::step_interact(...)
#>  2. │ └─recipes::add_step(...)
#>  3. ├─recipes::step_dummy(...)
#>  4. │ └─recipes::add_step(...)
#>  5. ├─recipes::recipe(...)
#>  6. └─recipes:::recipe.formula(...)
#>  7.   └─recipes:::form2args(formula, data, ...)
#>  8.     └─recipes:::inline_check(formula)
#>  9.       └─rlang::abort(...)
```

<sup>Created on 2023-01-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>


here is the example page that shows how formulas are set for lmer and nlme https://multilevelmod.tidymodels.org/articles/multilevelmod.html#linear-mixed-effects-via-lme

**Comment by yonicd** [comment]
this is also another way to run the model

``` r
library(tidymodels)
library(mmrm)
#> mmrm() registered as tidymodel linear_reg model

mmrm_spec <- linear_reg() |>
  set_engine("mmrm", method = 'Satterthwaite')

mmrm_wflow <- workflow() |>
  add_variables(outcomes = FEV1, predictors = c(RACE, ARMCD, AVISIT, USUBJID)) |> 
  add_model(mmrm_spec, formula = FEV1 ~ RACE + ARMCD * AVISIT + us(AVISIT | USUBJID))

mmrm_wflow |> 
  fit(data = fev_data)
#> ══ Workflow [trained] ══════════════════════════════════════════════════════════
#> Preprocessor: Variables
#> Model: linear_reg()
#> 
#> ── Preprocessor ────────────────────────────────────────────────────────────────
#> Outcomes: FEV1
#> Predictors: c(RACE, ARMCD, AVISIT, USUBJID)
#> 
#> ── Model ───────────────────────────────────────────────────────────────────────
#> mmrm fit
#> 
#> Formula:     FEV1 ~ RACE + ARMCD * AVISIT + us(AVISIT | USUBJID)
#> Data:        data (used 537 observations from 197 subjects with maximum 4 
#> timepoints)
#> Covariance:  unstructured (10 variance parameters)
#> Method:      REML
#> Deviance:    3387.373
#> 
#> Coefficients: 
#>                   (Intercept) RACEBlack or African American 
#>                   30.96769899                    1.50464863 
#>                     RACEWhite                      ARMCDTRT 
#>                    5.61309565                    3.77555734 
#>                    AVISITVIS2                    AVISITVIS3 
#>                    4.82858803                   10.33317002 
#>                    AVISITVIS4           ARMCDTRT:AVISITVIS2 
#>                   15.05255715                   -0.01737409 
#>           ARMCDTRT:AVISITVIS3           ARMCDTRT:AVISITVIS4 
#>                   -0.66753189                    0.63094392 
#> 
#> Model Inference Optimization:
#> Converged with code 0 and message: convergence: rel_reduction_of_f <= factr*epsmch
```

<sup>Created on 2023-01-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by danielinteractive** [comment]
thanks @yonicd for the review, very valuable!
From my outsider view adding a `step_*` method for the covariance structure specification could be good. 
(It would be really good if we don't add a silent argument to specify the covariance structure)

**Comment by yonicd** [comment]
@danielinteractive i thought about it more and steps_ doesnt make sense. It is data prep and not model assumptions. The only place that makes sense for this would be in the arguments that go into engine since it is mmrm specific. recipes should be an engine agnostic mechanism. It's usage is meant to be able to prep the data once and apply it to various model engines consistently. It basically returns the design matrix after applying transformations to the raw data.

From what i saw in my testing mmrm doesnt know what to do with formulas like `y ~ . ` which is common in other packages.

**Comment by danielinteractive** [comment]
@yonicd We discussed this and it is best to handle the `recipe` functionality in a follow-up issue, because it will require larger changes in this package. So we keep the scope contained in this PR/issue to the basic functionality.

**Comment by yonicd** [comment]
Do we want to hold off on merging until the prediction piece is in place? That will surely come up when people try to run it.

**Comment by dgkf** [comment]
@yonicd - I'm fine either way. Personally I don't use `tidymodels` enough to be able to gauge the usefulness of the interface without a `predict()` method. 

If you think it would be better to way, I'm fine with that.

**Comment by danielinteractive** [comment]
I would prefer to merge soon. The predict piece can take quite a while. We can file a separate follow up issue for that

**Comment by dgkf** [comment]
Yeah, on second thought I'd lean toward merging. We can aim to have the predict stuff before the next release, but this also includes some QOL improvements to resolve #221 and other related warnings.

File: NEWS.md
Comment (dgkf) [reply NEWS.md:6]: Sure, will make the change. 

In general, I think this is a great way of showing traceability of changes in case there are downstream effects or if users are curious about the decision-making process. The authorship bit to me is less critical, and not as widespread, but it is sort of a nice way to annotate code/feature ownership and elevate community contributors. 

It's seems to be quite a common practice among `r-lib` packages. Just picking the first three packages in my history: [remotes](https://github.com/r-lib/remotes/blob/main/NEWS.md), [covr](https://github.com/r-lib/covr/blob/main/NEWS.md), [dplyr](https://github.com/tidyverse/dplyr/blob/main/NEWS.md)

File: R/utils.R
Comment (danielinteractive) [inline R/utils.R:310]: Hm interesting! I am not sure about this though and needs more discussion. Can we please make this a separate issue / PR? So for this one please follow the standard formatting. Thanks

Comment (danielinteractive) [reply R/utils.R:310]: In particular, this would be one implementation way. Another would explicitly extend roxygen roclets, following https://github.com/r-lib/roxygen2/blob/main/vignettes/extending.Rmd. Plus I think it could be much better to have this done in a separate small extension package to roxygen to be able to use in all other packages we develop.

Comment (dgkf) [reply R/utils.R:310]: Yeah, [I've written roclets before](https://github.com/dgkf/testex) - that's definitely the more holistic way of solving this. In this case, I'm not sure it's worth the added dependency (even if just a dev dependency), nor am I entirely sold on the value-add of embedding type annotations for parameters in a dynamic language (for example, we have plenty of functions that require a specific class 

Having a function to at least keep `mmrm`'s particular formatting consistent seemed like the right middle ground.

I can revert these changes and we can discuss whether this is worth the effort to refactor from there.

Comment (danielinteractive) [reply R/utils.R]: Thanks Doug - good that you follow our conventions for now, and I suggest we discuss separately on this topic

File: R/zzz.R
Comment (danielinteractive) [inline R/zzz.R:31]:
```suggestion
#' @param message (`NULL` or `string`)\cr
```
this is our current convention :-)

File: R/zzz.R
Comment (danielinteractive) [inline R/zzz.R:69]: ok if so complicated we don't put in parentheses. basically I would just put numeric or just vector here and then explain in the argument description.

**Review by danielinteractive — APPROVED** [review]

**Pull request merged**
